pkg/scheduler/objects/node_collection.go

```diff
@@ -0,0 +1,245 @@
+	nc.RLock()
+	defer nc.RUnlock()
+	nref := nc.nodes[nodeID]
+	if nref == nil {
+		return nil
+	}
+	return nref.node
+}
+
+// Get the count of nodes
+func (nc *baseNodeCollection) GetNodeCount() int {
+	nc.RLock()
+	defer nc.RUnlock()
+	return len(nc.nodes)
+}
+
+// Return a list of nodes.
+func (nc *baseNodeCollection) GetNodes() []*Node {
+	nc.RLock()
+	defer nc.RUnlock()
+	nodes := make([]*Node, 0, len(nc.nodes))
+	for _, nref := range nc.nodes {
+		nodes = append(nodes, nref.node)
+	}
+	return nodes
+}
+
+// Return a list of schedulable nodes (optionally excluding reserved as well).
+func (nc *baseNodeCollection) GetSchedulableNodes(excludeReserved bool) []*Node {
+	nc.RLock()
+	defer nc.RUnlock()
+	nodes := make([]*Node, 0)
+	for _, nref := range nc.nodes {
+		node := nref.node
+		// filter out the nodes that are not scheduling
+		if !node.IsSchedulable() || (excludeReserved && node.IsReserved()) {
+			continue
+		}
+		nodes = append(nodes, node)
+	}
+	return nodes
+}
+
+// Create a node iterator for the schedulable nodes based on the policy set for this partition.
+// The iterator is nil if there are no schedulable nodes available.
+func (nc *baseNodeCollection) GetSchedulableNodeIterator() NodeIterator {
+	tree := nc.cloneActiveNodes()
+
+	length := tree.Len()
+	if length == 0 {
+		return nil
+	}
+
+	nodes := make([]*Node, 0, length)
+	tree.Ascend(func(item btree.Item) bool {
+		nodes = append(nodes, item.(nodeRef).node)
+		return true
+	})
+	return NewDefaultNodeIterator(nodes)
+}
+
+func (nc *baseNodeCollection) cloneActiveNodes() *btree.BTree {
+	nc.Lock()
+	defer nc.Unlock()
+
+	return nc.activeNodes.Clone()
+}
+
+// Sets the node sorting policy.
+func (nc *baseNodeCollection) SetNodeSortingPolicy(policy NodeSortingPolicy) {
+	nc.Lock()
+	defer nc.Unlock()
+	nc.nsp = policy
+
+	// activeNodes must be rebuilt since sort ordering is different
+	nc.activeNodes.Clear(false)
+	for _, nref := range nc.nodes {
+		node := nref.node
+		nref.nodeScore = nc.scoreNode(node)
+		if !node.IsSchedulable() || node.IsReserved() {
+			continue
+		}
+		nc.activeNodes.ReplaceOrInsert(*nref)
+	}
+}
+
+// Gets the node sorting policy.
+func (nc *baseNodeCollection) GetNodeSortingPolicy() NodeSortingPolicy {
+	nc.RLock()
+	defer nc.RUnlock()
+	return nc.nsp
+}
+
+// Callback method triggered when a node is updated.
+func (nc *baseNodeCollection) NodeUpdated(node *Node) {
+	nc.Lock()
+	defer nc.Unlock()
+
+	nref := nc.nodes[node.NodeID]
+	if nref == nil {
+		return
+	}
+
+	nc.activeNodes.Delete(*nref)
+	nref.nodeScore = nc.scoreNode(node)
+	if !node.IsSchedulable() || node.IsReserved() {
+		return
+	}
+	nc.activeNodes.ReplaceOrInsert(*nref)
+}
+
+// Create a new collection for the given partition.
+func NewNodeCollection(partition string) NodeCollection {
+	return &baseNodeCollection{
+		Partition:   partition,
+		nsp:         nil,
+		nodes:       make(map[string]*nodeRef),
+		activeNodes: btree.New(5),
```

Comment:
why degree=5? any reason behind this? Really want to understand more : )

Reply:
The degree determines the minimum number of items per tree node, and this seemed like a reasonable choice given upper bound of nodes would be maybe 10k at most. Consider a "full" tree of degree 5:

level 0 (root): 5 children
level 1: 5 * 5 children = 25
level 2: 25 * 5 children = 625
level 3: 625 * 5 children = 3125
level 4: 3125 * 5 children = 15625

Degree=5 means we can handle > 3k nodes with only depth=3, and > 15k with depth=4. At the same time, we only have max of 5 nodes to compare against at each level.

Reply:
Is this correct? 
```
level 2: 25 * 5 children = 625
```
shouldn't this be 125? does this mean we need depth=4 to handle 3k+ nodes, and depth=5 to handle 15k+ nodes?
what is the trade-off to have more items per tree node comparing to fewer items? I am asking because we may want to achieve the best results for less than 2000 nodes.

Reply:
Good catch. I think we want to optimize for larger cases though, as with smaller cluster sizes the number of containers we need to schedule per second is also less, so any algorithmic inefficiencies are mitigated.

I decided to do a little benchmarking to see if there's some clarity around an ideal degree value, and the results are interesting.

I ran 4 runs in a row for each of degree=3,4,5,6,7,8. I threw out the first run as a 'warm up' and then averaged the remaining 3 and calculated variance as well.

There's a lot of variance in the data, but surprisingly, all results show a net loss of throughput on 500 nodes vs. degree=5. So we appear to be handling that case optimally now.

At 1000 nodes, there's a modest (7-9%) gain in throughput for degree=6 and degree=7, and little change with 3, 4 or 8.
At 2000 nodes, we start to see significant gains (10%-18%) for degree=6 and degree=7, and again minimal change with 3, 4 or 8.
At 5000 nodes, the gains are again significant (15-16%) for degree=6,7, 11% for 8, with minimal change on 3 or 4.

It seems that degree=7 is optimal on >= 1000 nodes, with the largest gains of all the variants in best-case, average-case, and worst-case vs. 5. 

I also ran a few tests with very large (up to degree=10000), and the conclusion I come to is that the cost doesn't increase too much except at the very end (where we basically convert our B-Tree into a single node with a child list).

I'm updated the code to use 7, since that seems to be the optimal case, but the benchmarking seems to indicate that we're pretty safe regardless.